examples/cpp/PineAPPL.hpp

```diff
@@ -0,0 +1,257 @@
+    int get_int(const std::string& key) const { return pineappl_keyval_int(this->raw, key.c_str()); }
+    std::string get_string(const std::string& key) const { return pineappl_keyval_string(this->raw, key.c_str()); }
+    ///@}
+
+};
+
+/** @brief Entry in luminosity function  */
+struct LumiEntry {
+    /** @brief first parton id */
+    int32_t pid1;
+
+    /** @brief second parton id */
+    int32_t pid2;
+
+    /** @brief relative weight */
+    double weight;
+};
+
+/** @brief Luminosity function */
+struct Lumi {
+    /** @brief underlying raw object */
+    pineappl_lumi *raw;
+
+    /** @brief constructor */
+    Lumi(){ this->raw = pineappl_lumi_new(); }
+
+    /** @brief destructor */
+    ~Lumi(){ pineappl_lumi_delete(this->raw); }
+
+    /** @brief number of elements */
+    size_t count() const { return pineappl_lumi_count(this->raw); }
+
+    /**
+     * @brief add a luminosity function
+     * @param c luminosity function
+     */
+    void add(const std::vector<LumiEntry>& c) const {
+        size_t n = 0;
+        int32_t *pids = new int32_t;
+        double *weights = new double;
+        for (auto it = c.cbegin(); it != c.cend(); ++it, ++n) {
+            pids[2*n] = it->pid1;
+            pids[2*n+1] = it->pid2;
+            weights[n] = it->weight;
+        }
+        pineappl_lumi_add(this->raw, n, pids, weights);
+        delete pids;
+        delete weights;
+    }
+
+    /**
+     * @brief Returns the number of combinations of the luminosity function `lumi` for the specified entry.
+     * @param entry position in lumi
+     */
+    size_t combinations(size_t entry) const { return pineappl_lumi_combinations(this->raw, entry); }
+};
+
+/** @brief Coupling powers for each grid. */
+struct Order {
+    /** @brief Exponent of the strong coupling. */
+    uint32_t alphas;
+
+    /** @brief Exponent of the electromagnetic coupling. */
+    uint32_t alpha;
+
+    /** @brief Exponent of the logarithm of the scale factor of the renomalization scale. */
+    uint32_t logxir;
+
+    /** @brief Exponent of the logarithm of the scale factor of the factorization scale. */
+    uint32_t logxif;
+};
+
+class Grid {
+    /** @brief number of orders */
+    size_t n_orders = 0;
+
+    /** @brief number of bins */
+    size_t n_bins = 0;
+
+    /** @brief number of lumis */
+    size_t n_lumis = 0;
+
+public:
+
+    /** @brief underlying raw object */
+    pineappl_grid *raw;
+
+    /**
+     * @brief constructor
+     * @param lumi luminosity
+     * @param orders orders
+     * @param bins bins
+     * @param key_val additional informations
+     */
+    Grid(const Lumi& lumi, const std::vector<Order>& orders, const std::vector<double>& bins, const KeyVal& key_val) {
+        this->n_lumis = lumi.count();
+        // cast orders
+        uint32_t *raw_orders = new uint32_t[4 * orders.size()];
+        for (auto it = orders.cbegin(); it != orders.cend(); ++it, ++this->n_orders){
```

Comment:
Use a [range-for](https://en.cppreference.com/w/cpp/language/range-for) to iterate over `orders` and assign `this->n_orders` the correct value - this is much more readable than incrementing it up to the correct value.

Reply:
`n_orders` is now local and correct initialized - however, for the loop I need the index since `raw_orders` has more elements